File: study/lua/src/app/ui/component/CommonGroupsActiveTimeNode.lua
--组队活动时间显示控件
local CommonGroupsActiveTimeNode = class("CommonGroupsActiveTimeNode")

local EXPORTED_METHODS = {
    "updateUIOfStatic",
	"updateQinTomb",
	"setAddTouchCallback",
}

function CommonGroupsActiveTimeNode:ctor()
	self._target = nil
	self._showPopup = false
end

function CommonGroupsActiveTimeNode:_init()
	self._panelTouch = ccui.Helper:seekNodeByName(self._target, "PanelTouch")
	self._imageClockStatic = ccui.Helper:seekNodeByName(self._target, "ImageClockStatic")
	self._textTimeTitle = ccui.Helper:seekNodeByName(self._target, "TextTimeTitle")
	self._textLeftTime = ccui.Helper:seekNodeByName(self._target, "TextLeftTime")
	self._imageTouch = ccui.Helper:seekNodeByName(self._target, "ImageTouch")
	self._imageTouch:addClickEventListenerEx(handler(self, self._onClick))
	self._imagePopup = ccui.Helper:seekNodeByName(self._target, "ImagePopup")
	self._text1 = ccui.Helper:seekNodeByName(self._target, "Text1")
	self._text2 = ccui.Helper:seekNodeByName(self._target, "Text2")
	self._textTime1 = ccui.Helper:seekNodeByName(self._target, "TextTime1")
	self._textTime2 = ccui.Helper:seekNodeByName(self._target, "TextTime2")
	self._nodeTimeEffect = ccui.Helper:seekNodeByName(self._target, "NodeTimeEffect")

	self._panelAdd = ccui.Helper:seekNodeByName(self._target, "Panel_Add")
	self._panelAdd:addClickEventListener(handler(self, self._onAddTouch))
	self._onAddTouchCallback = nil

	self._text1:setString(Lang.get("groups_active_time_title"))
	self._text2:setString(Lang.get("groups_assist_time_title"))

	self._panelTouch:setContentSize(G_ResolutionManager:getDesignCCSize())
	self._panelTouch:setSwallowTouches(false)
	self._panelTouch:addClickEventListener(handler(self, self._onTouch)) --避免0.5秒间隔
end

function CommonGroupsActiveTimeNode:bind(target)
	self._target = target
    self:_init()
    cc.setmethods(target, self, EXPORTED_METHODS)
end

function CommonGroupsActiveTimeNode:unbind(target)
    cc.unsetmethods(target, EXPORTED_METHODS)
end

function CommonGroupsActiveTimeNode:_adjustTouchArea()
	local nodePos = self._target:convertToWorldSpaceAR(cc.p(0,0))
	local size = G_ResolutionManager:getDesignCCSize()
	local posx = size.width/2 - nodePos.x
	local posy = size.height/2 - nodePos.y
	self._panelTouch:setPosition(cc.p(posx, posy))
end

function CommonGroupsActiveTimeNode:updateUIOfStatic()
	self:_adjustTouchArea()

	self._imageClockStatic:loadTexture(Path.getQinTomb("img_qintomb_time01"))

	local leftTime = G_UserData:getBase():getGrave_left_sec()
	local assistLeftTime = G_UserData:getBase():getGrave_assist_sec()
	if leftTime > 0 then
		self._textTimeTitle:setString(Lang.get("groups_active_time_title"))
		local leftTimeStr = G_ServerTime:_secondToString(leftTime)
		self._textLeftTime:setString(leftTimeStr)
	else
		self._textTimeTitle:setString(Lang.get("groups_assist_time_title"))
		local leftTimeStr = G_ServerTime:_secondToString(assistLeftTime)
		self._textLeftTime:setString(leftTimeStr)
	end

	self._textTime1:setString(G_ServerTime:_secondToString(leftTime))
	self._textTime2:setString(G_ServerTime:_secondToString(assistLeftTime))

	self._imagePopup:setVisible(self._showPopup)
end

function CommonGroupsActiveTimeNode:_onClick()
	self._showPopup = not self._showPopup
	self._imagePopup:setVisible(self._showPopup)
end

function CommonGroupsActiveTimeNode:setAddTouchCallback (callback)
	self._onAddTouchCallback = callback
end

function CommonGroupsActiveTimeNode:_onAddTouch()
	if self._onAddTouchCallback then
		self:_onAddTouchCallback()
	end
end

function CommonGroupsActiveTimeNode:_onTouch()
	if self._showPopup then
		self._showPopup = false
		self._imagePopup:setVisible(self._showPopup)
	end
end

--更新秦皇陵
function CommonGroupsActiveTimeNode:updateQinTomb()
	local leftTime = G_UserData:getBase():getGrave_left_sec()
	local assistLeftTime = G_UserData:getBase():getGrave_assist_sec()

	self:updateUIOfStatic()
	if leftTime > 0 then
		self:_updateGraveLeftSec()
	elseif assistLeftTime > 0 then
		self:_updateGraveAssistSec()
	end
	
	self._panelAdd:setVisible(true)
end


function CommonGroupsActiveTimeNode:_updateGraveAssistSec( ... )
	local leftSec = G_UserData:getBase():getGrave_left_sec()
	local beginTime =  G_UserData:getBase():getGrave_begin_time()
	local assistLeftTime = G_UserData:getBase():getGrave_assist_sec()
	local assistBeginTime = G_UserData:getBase():getGrave_assist_begin_time()
	self._textTime2:stopAllActions()
	self._textTime1:stopAllActions()
	local function initUI()
		self._textTimeTitle:setString(Lang.get("groups_assist_time_title"))
		self._nodeTimeEffect:removeAllChildren()
		self._imageClockStatic:loadTexture(Path.getQinTomb("img_qintomb_time03"))
		G_EffectGfxMgr:createPlayGfx(self._nodeTimeEffect, "effect_xianqinhuangling_biaolan", nil ,true)

	end
	--协助时间处理
	if leftSec <=0 and assistLeftTime >0 then
		initUI()
		self:_playAssistLeftSec(self._textLeftTime)
		self:_playAssistLeftSec(self._textTime2)
		return true
	end
	return false
end



function CommonGroupsActiveTimeNode:_updateGraveLeftSec( ... )
	-- body
	self:_adjustTouchArea()

	local leftSec = G_UserData:getBase():getGrave_left_sec()
	local beginTime =  G_UserData:getBase():getGrave_begin_time()
	local assistLeftTime = G_UserData:getBase():getGrave_assist_sec()
	local assistBeginTime = G_UserData:getBase():getGrave_assist_begin_time()
	self._textTime2:stopAllActions()
	self._textTime1:stopAllActions()
	--协助时间处理
	local function initUI()
		self._textTimeTitle:setString(Lang.get("groups_active_time_title"))
		self._nodeTimeEffect:removeAllChildren()
		self._imageClockStatic:loadTexture(Path.getQinTomb("img_qintomb_time02"))
		G_EffectGfxMgr:createPlayGfx(self._nodeTimeEffect, "effect_xianqinhuangling_biao", nil ,true)
	end

	if beginTime > 0 then 
		initUI()
		self:_playGraveLeftSec(self._textLeftTime)
		self:_playGraveLeftSec(self._textTime1)
	end

end


--播放剩余倒计时
function CommonGroupsActiveTimeNode:_playGraveLeftSec(timeNode)

	local leftSec = G_UserData:getBase():getGrave_left_sec()
	local beginTime =  G_UserData:getBase():getGrave_begin_time()
	if beginTime == 0 then --暂停状态
		local endTime = leftSec + G_ServerTime:getTime()
		local leftTimeStr =  G_ServerTime:getLeftSecondsString(endTime, "00:00:00")
		timeNode:stopAllActions()
		timeNode:setString(leftTimeStr)
		return 
	end

	--timeNode:setString("00:00:00")
	local UIActionHelper = require("app.utils.UIActionHelper")
	local function timeUpdate()
		local endTime = leftSec + beginTime
		local leftTimeStr =  G_ServerTime:getLeftSecondsString(endTime, "00:00:00")
		local curTime = G_ServerTime:getTime()
		if  curTime > endTime then
			timeNode:stopAllActions()
			G_SignalManager:dispatch(SignalConst.EVENT_GRAVE_TIME_FINISH)
		else
			timeNode:setString(leftTimeStr)
		end
	end
	local action = UIActionHelper.createUpdateAction(function()
		timeUpdate()
	end, 0.5)

	timeNode:stopAllActions()
	timeNode:runAction(action)
end

--播放协助倒计时
function CommonGroupsActiveTimeNode:_playAssistLeftSec(timeNode)


	local leftSec = G_UserData:getBase():getGrave_assist_sec()
	local beginTime = G_UserData:getBase():getGrave_assist_begin_time()

	if beginTime == 0 then --暂停状态
		local endTime = leftSec + G_ServerTime:getTime()
		local leftTimeStr =  G_ServerTime:getLeftSecondsString(endTime, "00:00:00")
		timeNode:stopAllActions()
		timeNode:setString(leftTimeStr)
		return 
	end

	timeNode:setString("00:00:00")
	local UIActionHelper = require("app.utils.UIActionHelper")
	local function timeUpdate()
		local endTime = leftSec + beginTime
		local leftTimeStr =  G_ServerTime:getLeftSecondsString(endTime, "00:00:00")
		local curTime = G_ServerTime:getTime()
		if  curTime > endTime then
			timeNode:stopAllActions()
			G_SignalManager:dispatch(SignalConst.EVENT_GRAVE_TIME_FINISH)
		else
			timeNode:setString(leftTimeStr)
		end
	end
	local action = UIActionHelper.createUpdateAction(function()
		timeUpdate()
	end, 0.5)
	timeNode:runAction(action)
end
return CommonGroupsActiveTimeNode

File: study/ts/script/ui/component/CommonGroupsActiveTimeNode.ts
import { handler } from "../../utils/handler";
import { Lang } from "../../lang/Lang";
import { G_ResolutionManager, G_UserData, G_ServerTime, G_EffectGfxMgr, G_SignalManager } from "../../init";
import UIHelper from "../../utils/UIHelper";
import { Path } from "../../utils/Path";
import { SignalConst } from "../../const/SignalConst";
import UIActionHelper from "../../utils/UIActionHelper";

const { ccclass, property } = cc._decorator;

@ccclass
export default class CommonGroupsActiveTimeNode extends cc.Component {

    @property({ type: cc.Node, visible: true })
    _panelTouch: cc.Node = null;

    @property({ type: cc.Node, visible: true })
    _imageTouch: cc.Node = null;

    @property({ type: cc.Sprite, visible: true })
    _background: cc.Sprite = null;

    @property({ type: cc.Sprite, visible: true })
    _image_1: cc.Sprite = null;

    @property({ type: cc.Node, visible: true })
    _panel_Add: cc.Node = null;

    @property({ type: cc.Sprite, visible: true })
    _image_25: cc.Sprite = null;

    @property({ type: cc.Sprite, visible: true })
    _imageClockStatic: cc.Sprite = null;

    @property({ type: cc.Label, visible: true })
    _textTimeTitle: cc.Label = null;

    @property({ type: cc.Label, visible: true })
    _textLeftTime: cc.Label = null;

    @property({ type: cc.Node, visible: true })
    _nodeTimeEffect: cc.Node = null;

    @property({ type: cc.Sprite, visible: true })
    _imagePopup: cc.Sprite = null;

    @property({ type: cc.Label, visible: true })
    _text1: cc.Label = null;

    @property({ type: cc.Label, visible: true })
    _text2: cc.Label = null;

    @property({ type: cc.Label, visible: true })
    _textTime1: cc.Label = null;

    @property({ type: cc.Label, visible: true })
    _textTime2: cc.Label = null;

    private _showPopup = false;

    public onLoad() {
        this._imageTouch.on(cc.Node.EventType.TOUCH_START, handler(this, this._onClick))
        this._text1.string = (Lang.get('groups_active_time_title'));
        this._text2.string = (Lang.get('groups_assist_time_title'));
        this._panelTouch.setContentSize(G_ResolutionManager.getDesignCCSize());
        this._panelTouch.on(cc.Node.EventType.TOUCH_START, handler(this, this._onTouch));
        (this._panelTouch as any)._touchListener.setSwallowTouches(false);
    }

    public onEnable() {
    }

    public start() {
    }
    private _adjustTouchArea() {
        if (this.node.getComponent(cc.Widget) != null) {
            this.node.getComponent(cc.Widget).updateAlignment();
        }
        var nodePos = this.node.convertToWorldSpaceAR(cc.v2(0, 0));
        var size = G_ResolutionManager.getDesignCCSize();
        var posx = size.width / 2 - nodePos.x;
        var posy = size.height / 2 - nodePos.y;
        this._panelTouch.setPosition(posx, posy);
    }

    public updateUIOfStatic() {
        this._adjustTouchArea();
        UIHelper.loadTexture(this._imageClockStatic, Path.getQinTomb('img_qintomb_time01'));
        var leftTime = G_UserData.getBase().getGrave_left_sec();
        var assistLeftTime = G_UserData.getBase().getGrave_assist_sec();
        if (leftTime > 0) {
            this._textTimeTitle.string = (Lang.get('groups_active_time_title'));
            var leftTimeStr = G_ServerTime._secondToString(leftTime);
            this._textLeftTime.string = (leftTimeStr);
        } else {
            this._textTimeTitle.string = (Lang.get('groups_assist_time_title'));
            var leftTimeStr = G_ServerTime._secondToString(assistLeftTime);
            this._textLeftTime.string = (leftTimeStr);
        }
        this._textTime1.string = (G_ServerTime._secondToString(leftTime));
        this._textTime2.string = (G_ServerTime._secondToString(assistLeftTime));
        this._imagePopup.node.active = (this._showPopup);
    }

    private _onClick() {
        this._showPopup = !this._showPopup;
        this._imagePopup.node.active = (this._showPopup);
    }

    private _onTouch(event:cc.Event) {
        if (this._showPopup) {
            this._showPopup = false;
            this._imagePopup.node.active = (this._showPopup);
        }
    }

    public updateQinTomb() {
        var leftTime = G_UserData.getBase().getGrave_left_sec();
        var assistLeftTime = G_UserData.getBase().getGrave_assist_sec();
        this.updateUIOfStatic();
        if (leftTime > 0) {
            this._updateGraveLeftSec();
        } else if (assistLeftTime > 0) {
            this._updateGraveAssistSec();
        }
    }

    private _updateGraveAssistSec() {
        var leftSec = G_UserData.getBase().getGrave_left_sec();
        var beginTime = G_UserData.getBase().getGrave_begin_time();
        var assistLeftTime = G_UserData.getBase().getGrave_assist_sec();
        var assistBeginTime = G_UserData.getBase().getGrave_assist_begin_time();
        this._textTime2.node.stopAllActions();
        this._textTime1.node.stopAllActions();

        if (leftSec <= 0 && assistLeftTime > 0) {
            this._textTimeTitle.string = (Lang.get('groups_assist_time_title'));
            this._nodeTimeEffect.removeAllChildren();
            UIHelper.loadTexture(this._imageClockStatic, Path.getQinTomb('img_qintomb_time03'));
            G_EffectGfxMgr.createPlayGfx(this._nodeTimeEffect, 'effect_xianqinhuangling_biaolan', null, true);
            this._playAssistLeftSec(this._textLeftTime);
            this._playAssistLeftSec(this._textTime2);
            return true;
        }
        return false;
    }

    private _updateGraveLeftSec() {
        this._adjustTouchArea();
        var leftSec = G_UserData.getBase().getGrave_left_sec();
        var beginTime = G_UserData.getBase().getGrave_begin_time();
        var assistLeftTime = G_UserData.getBase().getGrave_assist_sec();
        var assistBeginTime = G_UserData.getBase().getGrave_assist_begin_time();
        this._textTime2.node.stopAllActions();
        this._textTime1.node.stopAllActions();

        if (beginTime > 0) {
            this._textTimeTitle.string = (Lang.get('groups_active_time_title'));
            this._nodeTimeEffect.removeAllChildren();
            UIHelper.loadTexture(this._imageClockStatic, Path.getQinTomb('img_qintomb_time02'));
            G_EffectGfxMgr.createPlayGfx(this._nodeTimeEffect, 'effect_xianqinhuangling_biao', null, true);
            this._playGraveLeftSec(this._textLeftTime);
            this._playGraveLeftSec(this._textTime1);
        }
    }

    private _playGraveLeftSec(timeNode: cc.Label) {
        var leftSec = G_UserData.getBase().getGrave_left_sec();
        var beginTime = G_UserData.getBase().getGrave_begin_time();
        if (beginTime == 0) {
            var endTime = leftSec + G_ServerTime.getTime();
            var leftTimeStr = G_ServerTime.getLeftSecondsString(endTime, '00:00:00');
            timeNode.node.stopAllActions();
            timeNode.string = (leftTimeStr);
            return;
        }
        timeNode.string = ('00:00:00');
        function timeUpdate() {
            var endTime = leftSec + beginTime;
            var leftTimeStr = G_ServerTime.getLeftSecondsString(endTime, '00:00:00');
            var curTime = G_ServerTime.getTime();
            if (curTime > endTime) {
                timeNode.node.stopAllActions();
                G_SignalManager.dispatch(SignalConst.EVENT_GRAVE_TIME_FINISH);
            } else {
                timeNode.string = (leftTimeStr);
            }
        }
        var action = UIActionHelper.createUpdateAction(function () {
            timeUpdate();
        }, 0.5);
        timeNode.node.runAction(action);
    }

    private _playAssistLeftSec(timeNode: cc.Label) {
        var leftSec = G_UserData.getBase().getGrave_assist_sec();
        var beginTime = G_UserData.getBase().getGrave_assist_begin_time();
        if (beginTime == 0) {
            var endTime = leftSec + G_ServerTime.getTime();
            var leftTimeStr = G_ServerTime.getLeftSecondsString(endTime, '00:00:00');
            timeNode.node.stopAllActions();
            timeNode.string = (leftTimeStr);
            return;
        }
        timeNode.string = ('00:00:00');
        function timeUpdate() {
            var endTime = leftSec + beginTime;
            var leftTimeStr = G_ServerTime.getLeftSecondsString(endTime, '00:00:00');
            var curTime = G_ServerTime.getTime();
            if (curTime > endTime) {
                timeNode.node.stopAllActions();
                G_SignalManager.dispatch(SignalConst.EVENT_GRAVE_TIME_FINISH);
            } else {
                timeNode.string = (leftTimeStr);
            }
        }
        var action = UIActionHelper.createUpdateAction(function () {
            timeUpdate();
        }, 0.5);
        timeNode.node.runAction(action);
    }
}